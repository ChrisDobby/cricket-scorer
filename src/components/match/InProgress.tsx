import * as React from 'react';
import { connect, Dispatch } from 'react-redux';
import { Team, State, Match, Innings, Batter, Bowler } from '../../domain';
import WithNavBar from '../WithNavBar';
import { StartInnings } from './StartInnings';
import { SelectBowler } from './SelectBowler';
import { BallEntry } from './BallEntry';
import * as actions from '../../actions/index';
import { BallFunctions } from './EntryPanel';

interface InProgressProps {
    match: Match;
    currentInnings?: Innings;
    currentBatter?: Batter;
    currentBowler?: Bowler;
    currentOverComplete?: boolean;
    startInnings?: (t: Team, b1: number, b2: number) => void;
    newBowler?: (b: number) => void;
    ballFunctions?: BallFunctions;
}

class InProgress extends React.Component<InProgressProps, {}> {
    render() {
        if (this.props.match && !this.props.currentInnings && this.props.startInnings) {
            return (
                <StartInnings
                    teams={[this.props.match.homeTeam, this.props.match.awayTeam]}
                    startInnings={this.props.startInnings}
                />
            );
        }

        if (this.props.currentInnings && !this.props.currentBowler && this.props.newBowler) {
            return (
                <SelectBowler
                    bowlingTeam={this.props.currentInnings.bowlingTeam}
                    selectBowler={this.props.newBowler}
                />);
        }

        if (this.props.currentInnings &&
            this.props.currentBatter &&
            this.props.currentBowler &&
            this.props.ballFunctions) {
            return (
                <BallEntry
                    innings={this.props.currentInnings}
                    batter={this.props.currentBatter}
                    bowler={this.props.currentBowler}
                    overComplete={!!this.props.currentOverComplete}
                    ballFunctions={this.props.ballFunctions}
                />
            );
        }

        return <div />;
    }
}

const mapStateToProps = (state: State) => ({
    match: state.inProgress.match,
    currentInnings: state.inProgress.currentInnings,
    currentBatter: state.inProgress.currentBatter,
    currentBowler: state.inProgress.currentBowler,
    currentOverComplete: state.inProgress.currentOverComplete,
});

const mapDispatchToProps = (dispatch: Dispatch<actions.InningsAction>) => ({
    startInnings: (team: Team, batter1Index: number, batter2Index: number) =>
        dispatch(actions.startInnings(team, batter1Index, batter2Index)),
    newBowler: (bowlerIndex: number) => dispatch(actions.newBowler(bowlerIndex)),
    ballFunctions: {
        dotBall: () => dispatch(actions.dotBall()),
        completeOver: () => dispatch(actions.completeOver()),
    },
});

export default connect(mapStateToProps, mapDispatchToProps)(WithNavBar(InProgress));
