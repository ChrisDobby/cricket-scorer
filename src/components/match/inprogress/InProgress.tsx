import * as React from 'react';
import { withRouter, RouteComponentProps } from 'react-router-dom';
import { Observer } from 'mobx-react';
import { History } from 'history';
import Paper from '@material-ui/core/Paper';
import { withStyles } from '@material-ui/core/styles';
import * as domain from '../../../domain';
import StartInnings from './StartInnings';
import SelectBowler from './SelectBowler';
import BallEntry from './BallEntry';
import Innings from '../../scorecard/Innings';
import SelectNewBatter from './SelectNewBatter';
import InningsComplete from './InningsComplete';
import CompleteMatch from '../CompleteMatch';
import EntryContainer from './EntryContainer';
import { bindMatchStorage } from '../../../stores/withMatchStorage';
import { getTeam } from '../../../match/utilities';
import calculateResult from '../../../match/calculateResult';
import getPlayers from '../../../match/getPlayers';
import battingMinutes from '../../../match/innings/battingMinutes';

const styles = (theme: any) => ({
    root: {
        ...theme.mixins.gutters(),
        paddingTop: theme.spacing.unit * 2,
        paddingBottom: theme.spacing.unit * 2,
        margin: '20px',
    },
});

type InProgressProps = RouteComponentProps<{}> & {
    inProgress: domain.InProgressMatch;
    storeMatch: (m: domain.InProgressMatch) => void;
    history: History;
    classes: any;
    userProfile: domain.Profile;
};

const InProgress = (props: InProgressProps) => {
    const bindStorage = bindMatchStorage(props.storeMatch, () => props.inProgress, () => props.userProfile.id);

    const ballFunctions = bindStorage({
        delivery: props.inProgress.delivery,
        completeOver: props.inProgress.completeOver,
        undoPreviousDelivery: props.inProgress.undoPreviousDelivery,
        completeInnings: props.inProgress.completeInnings,
        completeMatch: props.inProgress.completeMatch,
    });

    const disallowedPlayers = () =>
        typeof props.inProgress.previousBowler === 'undefined' ? [] : [props.inProgress.previousBowler.playerIndex];

    const previousBowlerFromEndIndex = () =>
        typeof props.inProgress.previousBowlerFromEnd === 'undefined'
            ? undefined
            : props.inProgress.previousBowlerFromEnd.playerIndex;

    const get = (innings: domain.Innings) => getPlayers(props.inProgress.match, innings);

    const calculateMinutes = (innings: domain.BattingInnings) =>
        battingMinutes(() => new Date().getTime())(innings, props.inProgress.match.breaks);

    return (
        <Paper className={props.classes.root}>
            <Observer
                render={() => (
                    <>
                        {!props.inProgress.currentInnings && !props.inProgress.provisionalMatchComplete && (
                            <StartInnings
                                teams={[props.inProgress.match.homeTeam, props.inProgress.match.awayTeam]}
                                startInnings={bindStorage(props.inProgress.startInnings)}
                                defaultBattingTeam={props.inProgress.nextBattingTeam}
                                canChangeBattingTeam={props.inProgress.canSelectBattingTeamForInnings}
                                maximumOvers={props.inProgress.match.config.oversPerSide}
                            />
                        )}
                        {props.inProgress.currentInnings &&
                            !props.inProgress.currentBowler &&
                            props.inProgress.provisionalInningsStatus === domain.InningsStatus.InProgress && (
                                <SelectBowler
                                    bowlingTeam={getTeam(
                                        props.inProgress.match,
                                        props.inProgress.currentInnings.bowlingTeam,
                                    )}
                                    selectBowler={bindStorage(props.inProgress.newBowler)}
                                    initiallySelected={previousBowlerFromEndIndex()}
                                    disallowedPlayers={disallowedPlayers()}
                                />
                            )}
                        {props.inProgress.newBatterRequired &&
                            props.inProgress.provisionalInningsStatus === domain.InningsStatus.InProgress && (
                                <SelectNewBatter
                                    batting={(props.inProgress.currentInnings as domain.Innings).batting}
                                    players={
                                        getTeam(
                                            props.inProgress.match,
                                            (props.inProgress.currentInnings as domain.Innings).battingTeam,
                                        ).players
                                    }
                                    batterSelected={bindStorage(props.inProgress.newBatter)}
                                />
                            )}
                        {props.inProgress.currentInnings && (
                            <>
                                {props.inProgress.currentBatter &&
                                    props.inProgress.currentBowler &&
                                    props.inProgress.currentOver &&
                                    !props.inProgress.newBatterRequired && (
                                        <>
                                            <BallEntry
                                                innings={props.inProgress.currentInnings}
                                                batter={props.inProgress.currentBatter}
                                                bowler={props.inProgress.currentBowler}
                                                overComplete={!!props.inProgress.currentOverComplete}
                                                currentOver={props.inProgress.currentOver}
                                                battingTeam={getTeam(
                                                    props.inProgress.match,
                                                    props.inProgress.currentInnings.battingTeam,
                                                )}
                                                bowlingTeam={getTeam(
                                                    props.inProgress.match,
                                                    props.inProgress.currentInnings.bowlingTeam,
                                                )}
                                                homeTeam={props.inProgress.match.homeTeam.name}
                                                awayTeam={props.inProgress.match.awayTeam.name}
                                                calculateResult={() => calculateResult(props.inProgress.match)}
                                                delivery={ballFunctions.delivery}
                                                completeOver={ballFunctions.completeOver}
                                            />
                                            <EntryContainer>
                                                <Innings
                                                    innings={props.inProgress.currentInnings}
                                                    getTeam={type => getTeam(props.inProgress.match, type)}
                                                    getBowlerAtIndex={
                                                        get(props.inProgress.currentInnings).getBowlerAtIndex
                                                    }
                                                    getFielderAtIndex={
                                                        get(props.inProgress.currentInnings).getFielderAtIndex
                                                    }
                                                    sameBowlerAndFielder={
                                                        get(props.inProgress.currentInnings).sameBowlerAndFielder
                                                    }
                                                    calculateMinutes={calculateMinutes}
                                                />
                                            </EntryContainer>
                                        </>
                                    )}
                                {typeof props.inProgress.provisionalInningsStatus !== 'undefined' &&
                                    props.inProgress.provisionalInningsStatus !== domain.InningsStatus.InProgress && (
                                        <InningsComplete
                                            status={props.inProgress.provisionalInningsStatus}
                                            battingTeam={
                                                getTeam(
                                                    props.inProgress.match,
                                                    props.inProgress.currentInnings.battingTeam,
                                                ).name
                                            }
                                            complete={() =>
                                                ballFunctions.completeInnings(props.inProgress.provisionalInningsStatus)
                                            }
                                            undoPrevious={ballFunctions.undoPreviousDelivery}
                                        />
                                    )}
                            </>
                        )}
                        {props.inProgress.provisionalMatchComplete && (
                            <CompleteMatch
                                homeTeam={props.inProgress.match.homeTeam.name}
                                awayTeam={props.inProgress.match.awayTeam.name}
                                complete={result => {
                                    ballFunctions.completeMatch(result);
                                    props.history.push('/scorecard');
                                }}
                                disallowCancel
                                cancel={() => {}}
                                calculateResult={() => calculateResult(props.inProgress.match)}
                            />
                        )}
                    </>
                )}
            />
        </Paper>
    );
};

export default withStyles(styles)(withRouter(InProgress));
