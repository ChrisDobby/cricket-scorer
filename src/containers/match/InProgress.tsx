import * as React from 'react';
import { inject } from 'mobx-react';
import { default as InProgressComponent } from '../../components/match/inprogress/InProgress';
import storeMatch from '../../storeMatch';
import WithInProgressStore from '../../components/WithInProgressStore';
import WithInProgressMatch from '../../components/WithInProgressMatch';
import WithMatchDrawer from '../../components/match/WithMatchDrawer';
import WithMatchActions from '../../components/match/WithMatchActions';
import PageContext from '../../context/PageContext';
import WithLoggedOutWarning from '../../components/match/WithLoggedOutWarning';
import WithMatchNetworkStatus from '../../components/WithMatchNetworkStatus';

const InProgress = WithMatchNetworkStatus(
    WithLoggedOutWarning((props: any) => {
        const { setOptions } = React.useContext(PageContext);
        React.useEffect(
            () =>
                setOptions({
                    openDrawer: props.openDrawer,
                    showHelp: true,
                }),
            [],
        );
        return (
            <InProgressComponent
                {...props}
                inProgress={props.inProgressMatchStore}
                storeMatch={storeMatch(props.isAuthenticated, props.inProgressMatchStore.setId)}
            />
        );
    }),
);

export default WithInProgressStore()(
    inject('inProgressMatchStore')(WithMatchActions(WithInProgressMatch(WithMatchDrawer(InProgress)))),
);
