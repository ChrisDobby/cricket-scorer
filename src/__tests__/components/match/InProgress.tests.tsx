import * as React from 'react';
import * as ReactTestRenderer from 'react-test-renderer';
import { StaticRouter } from 'react-router';
import InProgress from '../../../components/match/InProgress';
import { blankMatch, matchWithStartedInnings, matchWithOverReadyToComplete } from '../../testData/matches';


describe('InProgress', () => {
    const noMatchProgress = {
        match: undefined,
        currentInnings: undefined,
        currentBatter: undefined,
        currentBowler: undefined,
        previousBowler: undefined,
        previousBowlerFromEnd: undefined,
        currentOver: undefined,
        currentOverComplete: undefined,
        startInnings: jest.fn(),
        newBowler: jest.fn(),
        delivery: jest.fn(),
        completeOver: jest.fn(),
        flipBatters: jest.fn(),
        completeInnings: jest.fn(),
    };

    const notStartedMatchProgress = {
        ...noMatchProgress,
        match: blankMatch,
    };

    const matchWithStartedInningsProgress = {
        ...notStartedMatchProgress,
        match: matchWithStartedInnings,
        currentInnings: matchWithStartedInnings.innings[0],
    };

    const matchDuringOverProgress = {
        ...matchWithStartedInningsProgress,
        match: matchWithOverReadyToComplete,
        currentInnings: matchWithOverReadyToComplete.innings[0],
        currentBatter: matchWithOverReadyToComplete.innings[0].batting.batters[0],
        currentBowler: matchWithOverReadyToComplete.innings[0].bowlers[0],
        currentOver: {
            bowlingRuns: 0,
            wickets: 0,
            deliveries: [],
        },
    };

    const storage = {
        storeMatch: jest.fn(),
    };

    it('should render correctly when no match', () => {
        const inProgress = ReactTestRenderer.create(<InProgress inProgress={noMatchProgress} storage={storage} />);

        expect(inProgress.toJSON()).toMatchSnapshot();
    });

    it('should render correctly for match that has not started', () => {
        const inProgress = ReactTestRenderer.create(
            <InProgress
                inProgress={notStartedMatchProgress}
                storage={storage}
            />);

        expect(inProgress.toJSON()).toMatchSnapshot();
    });

    it('should render correctly for match with started innings', () => {
        const inProgress = ReactTestRenderer.create(
            <InProgress
                inProgress={matchWithStartedInningsProgress}
                storage={storage}
            />);

        expect(inProgress.toJSON()).toMatchSnapshot();
    });

    it('should render correctly when there is a previous bowler', () => {
        const withPreviousBowler = {
            ...matchWithStartedInningsProgress,
            previousBowler: matchWithOverReadyToComplete.innings[0].bowlers[0],
        };

        const inProgress = ReactTestRenderer.create(
            <StaticRouter context={{}}>
                <InProgress inProgress={withPreviousBowler} storage={storage} />
            </StaticRouter>);

        expect(inProgress.toJSON()).toMatchSnapshot();
    });

    it('should render correctly when there is a previous bowler from the current end', () => {
        const withPreviousBowlerFromEnd = {
            ...matchWithStartedInningsProgress,
            previousBowlerFromEnd: matchWithOverReadyToComplete.innings[0].bowlers[0],
        };

        const inProgress = ReactTestRenderer.create(
            <StaticRouter context={{}}>
                <InProgress
                    inProgress={withPreviousBowlerFromEnd}
                    storage={storage}
                />
            </StaticRouter>);

        expect(inProgress.toJSON()).toMatchSnapshot();
    });

    it('should render correctly for match during an over', () => {
        const inProgress = ReactTestRenderer.create(
            <StaticRouter context={{}}>
                <InProgress
                    inProgress={matchDuringOverProgress}
                    storage={storage}
                />
            </StaticRouter>);

        expect(inProgress.toJSON()).toMatchSnapshot();
    });
});
