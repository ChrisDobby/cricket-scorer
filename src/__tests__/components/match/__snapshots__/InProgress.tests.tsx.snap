// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InProgress should render correctly for match that hasn not started 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "height": "100vh",
    }
  }
>
  <nav
    className="navbar navbar-dark bg-primary"
  >
    <a
      className="navbar-brand"
      href="#"
    >
      Cricket scorer
    </a>
  </nav>
  <div
    style={
      Object {
        "flex": "1",
        "marginTop": "8px",
        "overflowX": "hidden",
        "overflowY": "auto",
      }
    }
  >
    <div>
      <div
        className="row"
      >
        <div
          className="d-none d-md-block d-lg-block col-2 col-lg-3"
        />
        <div
          className="col-12 col-md-8 col-lg-6"
        >
          <div
            style={
              Object {
                "marginBottom": "10px",
                "padding": "20px",
                "width": "100%",
              }
            }
          >
            <div
              className="row"
              style={
                Object {
                  "backgroundColor": "rgb(233, 236, 239)",
                  "paddingLeft": "15px",
                  "paddingTop": "4px",
                }
              }
            >
              <h4>
                Select batting team
              </h4>
            </div>
            <div
              className="form-check"
            >
              <input
                checked={false}
                className="form-check-input"
                id="teamRadio0"
                onChange={[Function]}
                type="radio"
                value={0}
              />
              <label
                className="form-check-label"
                htmlFor="teamRadio0"
              >
                Team 1
              </label>
            </div>
            <div
              className="form-check"
            >
              <input
                checked={false}
                className="form-check-input"
                id="teamRadio1"
                onChange={[Function]}
                type="radio"
                value={1}
              />
              <label
                className="form-check-label"
                htmlFor="teamRadio1"
              >
                Team 2
              </label>
            </div>
          </div>
        </div>
      </div>
      <button
        className="btn btn-success"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "borderRadius": "25px",
            "bottom": "10px",
            "boxShadow": "0 4px 6px 0 rgba(0,0,0,.14), 0 4px 5px rgba(0,0,0,-1)",
            "fontSize": "18px",
            "height": "50px",
            "lineHeight": "1.33",
            "padding": "10px 16px",
            "position": "fixed",
            "right": "10px",
            "width": "50px",
            "zIndex": 1049,
          }
        }
        type="button"
      >
        <i
          className="fa fa-check"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`InProgress should render correctly for match with started innings 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "height": "100vh",
    }
  }
>
  <nav
    className="navbar navbar-dark bg-primary"
  >
    <a
      className="navbar-brand"
      href="#"
    >
      Cricket scorer
    </a>
  </nav>
  <div
    style={
      Object {
        "flex": "1",
        "marginTop": "8px",
        "overflowX": "hidden",
        "overflowY": "auto",
      }
    }
  >
    <div>
      <div
        className="row"
      >
        <div
          className="d-none d-md-block d-lg-block col-2 col-lg-3"
        />
        <div
          className="col-12 col-md-8 col-lg-6"
        >
          <div
            style={
              Object {
                "marginBottom": "10px",
                "padding": "20px",
                "width": "100%",
              }
            }
          >
            <div
              className="row"
              style={
                Object {
                  "backgroundColor": "rgb(233, 236, 239)",
                  "paddingLeft": "15px",
                  "paddingTop": "4px",
                }
              }
            >
              <h4>
                Select bowler
              </h4>
            </div>
            <div
              className="row bowler-row"
              onClick={[Function]}
              style={
                Object {
                  "cursor": "pointer",
                }
              }
            >
              <div
                style={
                  Object {
                    "color": "#007bff",
                    "float": "left",
                    "textAlign": "center",
                    "width": "30px",
                  }
                }
              />
              <h6>
                Player 12
              </h6>
            </div>
            <div
              className="row bowler-row"
              onClick={[Function]}
              style={
                Object {
                  "cursor": "pointer",
                }
              }
            >
              <div
                style={
                  Object {
                    "color": "#007bff",
                    "float": "left",
                    "textAlign": "center",
                    "width": "30px",
                  }
                }
              />
              <h6>
                Player 13
              </h6>
            </div>
            <div
              className="row bowler-row"
              onClick={[Function]}
              style={
                Object {
                  "cursor": "pointer",
                }
              }
            >
              <div
                style={
                  Object {
                    "color": "#007bff",
                    "float": "left",
                    "textAlign": "center",
                    "width": "30px",
                  }
                }
              />
              <h6>
                Player 14
              </h6>
            </div>
            <div
              className="row bowler-row"
              onClick={[Function]}
              style={
                Object {
                  "cursor": "pointer",
                }
              }
            >
              <div
                style={
                  Object {
                    "color": "#007bff",
                    "float": "left",
                    "textAlign": "center",
                    "width": "30px",
                  }
                }
              />
              <h6>
                Player 15
              </h6>
            </div>
            <div
              className="row bowler-row"
              onClick={[Function]}
              style={
                Object {
                  "cursor": "pointer",
                }
              }
            >
              <div
                style={
                  Object {
                    "color": "#007bff",
                    "float": "left",
                    "textAlign": "center",
                    "width": "30px",
                  }
                }
              />
              <h6>
                Player 16
              </h6>
            </div>
            <div
              className="row bowler-row"
              onClick={[Function]}
              style={
                Object {
                  "cursor": "pointer",
                }
              }
            >
              <div
                style={
                  Object {
                    "color": "#007bff",
                    "float": "left",
                    "textAlign": "center",
                    "width": "30px",
                  }
                }
              />
              <h6>
                Player 17
              </h6>
            </div>
            <div
              className="row bowler-row"
              onClick={[Function]}
              style={
                Object {
                  "cursor": "pointer",
                }
              }
            >
              <div
                style={
                  Object {
                    "color": "#007bff",
                    "float": "left",
                    "textAlign": "center",
                    "width": "30px",
                  }
                }
              />
              <h6>
                Player 18
              </h6>
            </div>
            <div
              className="row bowler-row"
              onClick={[Function]}
              style={
                Object {
                  "cursor": "pointer",
                }
              }
            >
              <div
                style={
                  Object {
                    "color": "#007bff",
                    "float": "left",
                    "textAlign": "center",
                    "width": "30px",
                  }
                }
              />
              <h6>
                Player 19
              </h6>
            </div>
            <div
              className="row bowler-row"
              onClick={[Function]}
              style={
                Object {
                  "cursor": "pointer",
                }
              }
            >
              <div
                style={
                  Object {
                    "color": "#007bff",
                    "float": "left",
                    "textAlign": "center",
                    "width": "30px",
                  }
                }
              />
              <h6>
                Player 20
              </h6>
            </div>
            <div
              className="row bowler-row"
              onClick={[Function]}
              style={
                Object {
                  "cursor": "pointer",
                }
              }
            >
              <div
                style={
                  Object {
                    "color": "#007bff",
                    "float": "left",
                    "textAlign": "center",
                    "width": "30px",
                  }
                }
              />
              <h6>
                Player 21
              </h6>
            </div>
            <div
              className="row bowler-row"
              onClick={[Function]}
              style={
                Object {
                  "cursor": "pointer",
                }
              }
            >
              <div
                style={
                  Object {
                    "color": "#007bff",
                    "float": "left",
                    "textAlign": "center",
                    "width": "30px",
                  }
                }
              />
              <h6>
                Player 22
              </h6>
            </div>
          </div>
        </div>
      </div>
      <button
        className="btn btn-success"
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "borderRadius": "25px",
            "bottom": "10px",
            "boxShadow": "0 4px 6px 0 rgba(0,0,0,.14), 0 4px 5px rgba(0,0,0,-1)",
            "fontSize": "18px",
            "height": "50px",
            "lineHeight": "1.33",
            "padding": "10px 16px",
            "position": "fixed",
            "right": "10px",
            "width": "50px",
            "zIndex": 1049,
          }
        }
        type="button"
      >
        <i
          className="fa fa-check"
        />
      </button>
    </div>
  </div>
</div>
`;
